### DISCUSSION TITLE
Can I use staticx instead of compiling on an other distro?

### DISCUSSION THREAD

**Discussion by fwdekker** [pyinstaller/pyinstaller#7362] [Help] [answered]
(Edit: The full answer is spread over the marked answer and the comment thread below it.)

[The documentation states](https://pyinstaller.org/en/stable/usage.html#making-gnu-linux-apps-forward-compatible) that to make my Linux builds forwards-compatible, "the solution is to always build your app on the _oldest_ version of GNU/Linux you mean to support." However, in [discussion 5669](https://github.com/pyinstaller/pyinstaller/discussions/5669) it is mentioned that it's also possible to post-process my executable with `staticx`. I tried using `staticx` and it worked fine for me when running on an older Linux version than where I built it.

But I'm absolutely not well-versed with distributing libraries like this, so I have a few questions:
* How does this method compare to compiling on an older version, and are there any caveats I should be aware of? 
* Is there a reason `staticx` is not mentioned in the docs anywhere?

**Comment by rokm** [comment, marked as answer]
Trying to deploy binary software across different linux distributions (and even versions thereof) is full of caveats that come from unstable userspace ABI (i.e., incompatibilities with system-provided shared libs, including but not limited to glibc). Whether it works or not largely depends on how complex your program is in term of non-pure-python packages used and what shared libraries their extensions are using. And of course the differences between the build and target distribution.

> How does this method compare to compiling on an older version, and are there any caveats I should be aware of?

One caveat is that you're limited to onefile builds. Whether this is a consideration or not depends on the size of your program. Also, onefile builds will likely not work out-of-the-box on distributions that mount their /tmp with noexec.

There's also problem with shared libraries that have RUNPATH set, but those probably aren't common (and if you had this problem, you would know).

> Is there a reason `staticx` is not mentioned in the docs anywhere?

Not really. If you think it should be, feel free to add a paragraph or two.

**Reply by bwoodsend** [reply]
> There's also problem with shared libraries that have RUNPATH set, but those probably aren't common (and if you had this problem, you would know).

It's common if you use pyenv or the Python environments provided by GitHub Actions.

**Reply by fwdekker** [reply]
Thanks for the replies. I'm not too well-versed in this subject, so please bear with me ^^'

> incompatibilities with system-provided shared libs, including but not limited to glibc

But in a `staticx`'ed one-file build, that shouldn't happen, right, because in such a build the relevant libraries will be inside the build? Or is it still possible for there to be conflicts?

> onefile builds will likely not work out-of-the-box on distributions that mount their /tmp with noexec

That makes sense. I suppose that's a sacrifice I'm willing to make, since `apt` by default assumes `/tmp` is not mounted with `noexec` (as far as I could gather).

**Reply by rokm** [reply]
> > incompatibilities with system-provided shared libs, including but not limited to glibc
> 
> But in a `staticx`'ed one-file build, that shouldn't happen, right, because in such a build the relevant libraries will be inside the build? Or is it still possible for there to be conflicts?

If all shared libraries that your application and its dependencies need are collected in the frozen application bundle, then no, it should not happen.

I was referring to situations when you use something complex, like a UI framework (e.g., Qt), and end up depending on low-level system libraries (like X11 libs, mesa, DRI drivers, etc.) that are not collected. But I imagine such cases automatically preclude use of `staticx`, because those system-installed dependencies might not be found due to `staticx` manipulating the library search path.

Like I said, it depends, and if you want a more definite answer, you'll have to provide more details about what 3rd party python packages your code is actually using.

**Reply by fwdekker** [reply]
Thank you for the explanation! That mostly clears up my worries, and confirms I mostly understood your previous comment ^^

> But I imagine such cases automatically preclude use of `staticx`, because those system-installed dependencies might not be found due to `staticx` manipulating the library search path.

So if I understand you correctly, if I depend on Qt, and then use `staticx` on my one-file, the result will not work on _any_ system, not even my own, because the resulting executable won't search for shared libraries outside of the one in my distribution? If so, that's a relief of sorts :P

> if you want a more definite answer, you'll have to provide more details about what 3rd party python packages your code is actually using

The runtime dependencies of my project are [listed in `common.txt`](https://github.com/FWDekker/facemation/blob/a484b1352f2779f86cf94361f4de37ee9e2f6fc0/requirements/common.txt) (and for building it additionally depends on `pip-licenses`, `pyinstaller`, and `staticx`). Most dependencies are, I think, quite high-level (e.g. `natsort`, `tqdm`, `typeguard`); I suppose the lowest-level dependency is [`dlib`](http://dlib.net/)? I mean, that's the only one which requires build tools (`cmake` and `g++`) to be available during installation.

**Reply by rokm** [reply]
Hmm, it actually looks like `PySide6` and `PySide2`  use `DT_RUNPATH`, so `staticx` refuses to process onefile builds using them.

---

Also, `staticx` does seem to collect the libraries that PyInstaller does not, as evident from `psutil`-based dump of loaded libs in `staticx`-compiled application:
<details><summary>loaded libs</summary>

```
/tmp/_MEIT9G4wn/lib-dynload/resource.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/psutil/_psutil_posix.abi3.so
/tmp/_MEIT9G4wn/psutil/_psutil_linux.abi3.so
/tmp/_MEIT9G4wn/lib-dynload/grp.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/liblzma.so.5
/tmp/_MEIT9G4wn/lib-dynload/_lzma.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/libbz2.so.1
/tmp/_MEIT9G4wn/lib-dynload/_bz2.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/libquadmath.so.0
/tmp/_MEIT9G4wn/libgfortran.so.5
/tmp/staticx-CdeoBg/libxcb.so.1.1.0
/tmp/staticx-CdeoBg/libstdc++.so.6.0.30
/tmp/_MEIT9G4wn/libsqlite3.so.0
/tmp/_MEIT9G4wn/libopenblasp.so.0
/tmp/_MEIT9G4wn/libXau.so.6
/tmp/_MEIT9G4wn/libjpeg.so.62
/tmp/_MEIT9G4wn/libpng16.so.16
/tmp/_MEIT9G4wn/libXext.so.6
/tmp/_MEIT9G4wn/libX11.so.6
/tmp/_MEIT9G4wn/_dlib_pybind11.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/numpy/random/_generator.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/numpy/random/_sfc64.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/numpy/random/_pcg64.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/numpy/random/_philox.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/numpy/random/_mt19937.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/numpy/random/_bounded_integers.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/lib-dynload/_random.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/lib-dynload/_sha512.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/lib-dynload/_sha3.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/lib-dynload/_blake2.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/libcrypto.so.3
/tmp/_MEIT9G4wn/lib-dynload/_hashlib.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/numpy/random/_common.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/numpy/random/bit_generator.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/numpy/random/mtrand.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/numpy/fft/_pocketfft_internal.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/libquadmath-96973f99.so.0.0.0
/tmp/_MEIT9G4wn/lib-dynload/binascii.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/numpy/linalg/_umath_linalg.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/libgfortran-040039e1.so.5.0.0
/tmp/_MEIT9G4wn/numpy/core/_multiarray_tests.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/lib-dynload/_contextvars.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/lib-dynload/_datetime.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/libgcc_s.so.1
/tmp/_MEIT9G4wn/libopenblas64_p-r0-15028c96.3.21.so
/tmp/_MEIT9G4wn/numpy/core/_multiarray_umath.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/lib-dynload/_json.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/lib-dynload/_posixsubprocess.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/lib-dynload/array.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/lib-dynload/select.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/lib-dynload/math.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/lib-dynload/_socket.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/lib-dynload/_pickle.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/libffi.so.8
/tmp/_MEIT9G4wn/lib-dynload/_bisect.cpython-38-x86_64-linux-gnu.so
/usr/lib/locale/locale-archive
/tmp/_MEIT9G4wn/lib-dynload/_heapq.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/lib-dynload/zlib.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/lib-dynload/_struct.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/lib-dynload/_opcode.cpython-38-x86_64-linux-gnu.so
/tmp/_MEIT9G4wn/lib-dynload/_ctypes.cpython-38-x86_64-linux-gnu.so
/usr/lib64/gconv/gconv-modules.cache
/tmp/staticx-CdeoBg/libm.so.6
/tmp/_MEIT9G4wn/libpython3.8.so.1.0
/tmp/staticx-CdeoBg/libresolv.so.2
/tmp/staticx-CdeoBg/libnss_files.so.2
/tmp/staticx-CdeoBg/libnss_dns.so.2
/tmp/staticx-CdeoBg/libc.so.6
/tmp/staticx-CdeoBg/libpthread.so.0
/tmp/staticx-CdeoBg/libz.so.1.2.12
/tmp/staticx-CdeoBg/libdl.so.2
/tmp/staticx-CdeoBg/libnssfix.so
/tmp/staticx-CdeoBg/ld-linux-x86-64.so.2
``` 

</details>

Stuff in `/tmp/_MEIT9G4wn` is PyInstaller's, while stuff in `/tmp/staticx-CdeoBg` is `staticx`. For example, we explicitly [exclude libxcb.so.1.1.0](https://github.com/pyinstaller/pyinstaller/blob/2653893de7f6bfe05be30f9a60923987094e9091/PyInstaller/depend/dylib.py#L216), but for the purpose of the test I have also explicitly excluded `libstdc++.so.6`.

So my earlier comment on libraries not collected by PyInstaller not being found is wrong, because they might still be collected by `staticx` itself.

**Reply by fwdekker** [reply]
So essentially `staticx` also includes all the low-level system libraries, so there shouldn't ever be any incompatibilities; but if PyInstaller added something using `DT_RUNPATH` then `staticx` fails?

**Reply by rokm** [reply]
> So essentially staticx also includes all the low-level system libraries,

Yes. I *think* it checks the link-time dependencies of PyInstaller-collected shared libs (and checks for invalid RUNPATH at the same time), and then checks which of those are also PyInstaller-collected, and then collects the remaining ones. But I've never really looked that deep into its operation, so take what I say with grain of salt.

>  so there shouldn't ever be any incompatibilities;

Well, that's a bit strong statement (what with the only certainties in life being death and taxes...), but I think for the set of dependencies you listed, it should be fairly accurate. 

The non-pure-python packages in your case are `numpy`, `Pillow`, and `dlib`, and none of those link back to something that would be seem outright problematic (like the UI framework libs, or fontconfig/freetype, or the low-level graphics driver stack libraries).

> but if PyInstaller added something using DT_RUNPATH then staticx fails?

Yes, and thankfully that rules out `PySide2`/`PyQt5`/`PySide6`/`PyQt6`.

**Reply by fwdekker** [reply]
Awesome! That pretty much clears up the worries I had. Thank you for taking the time to answer my questions :smile:
